monix-eval/shared/src/main/scala/monix/eval/Task.scala

```diff
@@ -314,6 +314,21 @@ sealed abstract class Task[+A] extends Serializable { self =>
   def runAsync(implicit s: Scheduler): CancelableFuture[A] =
     TaskRunLoop.startAsFuture(this, s)
 
+  /** run the Task with LocalContext propagation.
+    *
+    * @param s is an injected [[monix.execution.Scheduler Scheduler]]
+    *        that gets used whenever asynchronous boundaries are needed
+    *        when evaluating the task
+    *
+    * @return a [[monix.execution.CancelableFuture CancelableFuture]]
+    *         that can be used to extract the result or to cancel
+    *         a running task.
+    */
+  def runAsyncTraced(implicit s: Scheduler): CancelableFuture[A] = {
```

Comment:
I like this `runAsyncTraced`, however we need to ensure that it works well, as I have doubts.

The problem with the run-loop is that within `Async` tasks users can restart that run-loop in however way they want. And we do that in our own implementation. For example:

```scala
Task.mapBoth(task1, task2) { (a, b) => a + b }
```

Within `mapBoth` we are starting `task1` and `task2` with the normal `runAsync`. Now indeed, according to your implementation, the local context should be restored at some point, probably in that mapping logic.

But will users expect for `task1` and `task2` to not receive that context on execution as well?
The other possibility would have been to introduce an extra option in `Task.Options`.

So what do you think?

Reply:
@alexandru you presenting these cases is useful so I can start getting more tests know if this works. I can start testing this case and evaluate.

My first concern about the proposal was about this type of cases. Is it something that we want to support for every constructor of a Task or leave a few ones to construct traced tasks? An additional option in `Task.Options` sounds reasonable if we want to extend this to all the Task API. I wanted to do the TaskTracedRunLoop to not force the tracing as default but to keep it optional, I assume that is still the intention.

Reply:
The only thing I care about is the performance impact for when the option is turned off.

When the option is turned off, I assume that's going to simply involve one or two `if` expressions being rejected (for reading and restoring that context respectively on async boundaries).

So if you can do a benchmark to see the impact of such if statements, that would be great.
We have a `benchmarks` sub-project, add a file in `benchmarks/next`, see the other files there because it's not hard to do it.

I like your approach, but I fear gotchas, plus there's always the issue of code duplication, which isn't bad, but then you have to retest all of those branches and it's difficult. So lets do an informed decision here - if it has a performance impact, we'll do it like you did it here, otherwise we simply modify the current run-loop implementation with the needed if statements.